core/src/page_id.rs

```diff
@@ -13,13 +30,28 @@ pub const ROOT_PAGE_ID: [u8; 32] = [0; 32];
 
 const MAX_CHILD_INDEX: u8 = (1 << DEPTH) - 1;
 
-pub fn child_page_id(mut page_id: PageId, child_index: u8) -> PageId {
-    // TODO: proper error handling
-    assert!(child_index & 0b11000000 == 0);
+/// Construct the Child PageId given the previous PageId and the child index.
+///
+/// Child index must be a 6 bit integer, two most significant bits must be zero.
+pub fn child_page_id(mut page_id: PageId, child_index: u8) -> Result<PageId, ChildPageIdError> {
+    if child_index > MAX_CHILD_INDEX {
+        return Err(ChildPageIdError::InvalidChildIndex);
+    }
 
     let page_id_bits = page_id.view_bits_mut::<Msb0>();
 
-    if child_index ^ MAX_CHILD_INDEX == 0 {
+    // The PageId of the first page for a layer I is:
+    // sum_i(2^(6 * i))
+    // where i goes from 0 to I - 1
+    //
+    // So the first page in the last layer has in position 246
+    // the last bit set to one. Every PageId with a 1 after that
+    // position will lead to a PageOverflow.
+    if dbg!(page_id_bits.leading_zeros()) <= 9 {
```

Comment:
comment does not explain the 9 - can't say I get it.
The comment here could be improved, and the logic may be wrong as well.

I see something like an argument that:
  1. The lowest page ID at the bottom of the page tree has ID `sum_i(2^(6 * i))`
  2. Therefore any page with an ID `>=` than this value cannot have children

But that is not what is implemented. Also, unless I am mistaken the lowest page ID at the bottom of the page tree equals `sum(2^(6k) for k 0..=42)` and this has bit 252=1.

What seems to be implemented is actually a different argument:
1. The highest page ID at the 2nd to last layer is equal to `sum(2^(1+6k) for k 1..=41` and this has bit 247 set as a result.
2. Any page ID which has a bit higher than 247 set is definitely in the last layer.
3. As above, the smallest last-layer page ID starts with bit 252 set, so this is exhaustive

But this gives me the check `leading_zeros < 9`, not `<= 9` because `= 9` means that bit 247 is set. So is the code here inaccurate or am I missing something?

In any case, we could also just as easily simplify the logic and have `leading_zeros < 4` as this means that bit 252 is set.

The comment gave me none of this. It could have been much more helpful.

Reply:
```rust
    #[test]
    fn last_page_id_test() {
        let levels = 42;

        // min page
        let mut page_id = ROOT_PAGE_ID;

        for i in 0..42 {
            page_id = child_page_id(page_id, 0).unwrap();
        }

        // max page
        let mut page_id = ROOT_PAGE_ID;

        for i in 0..42 {
            page_id = child_page_id(page_id, 1 << 6).unwrap()
        }
    }
```

fails against this diff, btw

Reply:
Sorry for the poor explanation.

You're totally right on how the layer before the last has the same bit in position 246 set to one, thanks for letting me notice that!

I mixed a couple of basic binary tree logic with the 64-ary tree we are dealing with. Within a layer, the most significant bit set to one can change (this cannot happen in a full binary tree), so working only with leading zeros is not enough.

However, `lowest(n) = sum_i(2^(1 + 6k)) for k 0..=n` doesn't appear correct to me. For example, `lowest(1)` would be `sum_i(2^(1 + 6k)) = 2 + 2^7 != 1`. Instead, what I came up with is:
![image](https://github.com/thrumdev/nomt/assets/79002163/7e77145d-396f-423a-a6eb-719a4750f855)

The formula does not work for layer 0, only between 1 and 42.

Another important thing to notice is that the page tree is made up of 43 layers, not 42. This is because there are 42 child indices for each KeyPath, so starting from layer zero (root) means arriving at layer 42.

So, given 42 as the index of the last layer of the page tree, the most significant bit set to one in layer 42 is `6 * 41 = 246`.

As you described, `lowest(42) = highest(41) + 1`, so bit 246 is set to one in both, and the solution I provided cannot be correct.

>The best actual solution I can think of is to reject every value with 9 leading zeroes or fewer unless it is bitwise less than `lowest(41)`.

I like this proposed solution, with a small variation. We know that the most significant bit set to one in `highest(42) = lowest(43) - 1` (even if `lowest(43)` doesn't really exist) is `252`. So, if there are less than or equal to 4 leading zeroes, there will be a direct page overflow; if there are between 5 and 9 leading zeroes, bitwise comparison to `lowest(42)` is necessary.

Reply:
> Another important thing to notice is that the page tree is made up of 43 layers, not 42. This is because there are 42 child indices for each KeyPath, so starting from layer zero (root) means arriving at layer 42.

sure, this depends how you look at it. I was counting from 0 as the layer below the root. There are 42 layers beyond the root which gives you a max index of 41 in that scheme. It' s a matter of definitions rather than truth.

> For example, lowest(1) would be sum_i(2^(1 + 6k)) = 2 + 2^7 != 1.

Yeah, I had this wrong. So I think the general function we are looking for is 

`path_n(depth) = sum_i(n * 64^i) i in 0..depth` 

I'll use your 1-counting which treats the depth zero as root.
This gives path_n(0) = 0 for all n.
path_1(1) = 1 * 1
path_64(1) = 64 * 1

so on.

> I like this proposed solution, with a small variation. We know that the most significant bit set to one in highest(42) = lowest(43) - 1 (even if lowest(43) doesn't really exist) is 252. So, if there are less than or equal to 3 leading zeroes, there will be a direct page overflow; if there are between 4 and 9 leading zeroes, bitwise comparison to lowest(42) is necessary.

This is just overcomplicating the solution. If a value has fewer leading zeroes than `highest(41)` then it definitely is in layer 42 and would overflow. If it has the same amount of leading zeroes it may also be in layer 2 and needs to be lexicographically compared. I don't see any benefit to adding another branch for 0-3 leading zeroes, except that we will overflow the container and this is different. But page IDs are logically 250 bits and we don't need to differentiate these cases.

Reply:
> This is just overcomplicating the solution

totally agree, thinking without coding made me overthinking about the problem